src/linalg.jl

```diff
@@ -92,8 +81,115 @@ end
     @invcheckoff for multivec_row in 1:size(X.parent, 2)
         for col in 1:size(A, 2)
             @inbounds for k in nzrange(A, col)
-                C[multivec_row, col] += X.parent[A.rowval[k], multivec_row]*A.nzval[k]*α
+                @routine begin
+                    anc ← zero(eltype(A))
+                    anc += X.parent[A.rowval[k], multivec_row]' * A.nzval[k]
+                end
+                C[multivec_row, col] += anc * α
+                ~@routine
+            end
+        end
+    end
+end
+
+@i function imul!(C::StridedVecOrMat, X::DenseMatrixUnion, xA::Adjoint{<:Any,<:AbstractSparseMatrix}, α::Number, β::Number)
+    @safe size(X, 2) == size(xA.parent, 2) || throw(DimensionMismatch())
+    @safe size(X, 1) == size(C, 1) || throw(DimensionMismatch())
+    @safe size(xA.parent, 1) == size(C, 2) || throw(DimensionMismatch())
+    if (β != 1, ~)
+        @safe error("only β = 1 is supported, got β = $(β).")
+    end
+    @invcheckoff for col in 1:size(xA.parent, 2)
+        @inbounds for k in nzrange(xA.parent, col)
+                        @routine begin
+                            anc1 ← zero(eltype(xA.parent))
+                            anc1 += (xA.parent.nzval[k])'*α
+                        end
+            @simd for multivec_row in 1:size(X,1)
+                C[multivec_row, xA.parent.rowval[k]] += X[multivec_row, col] * anc1
+            end
+            ~@routine
+        end
+    end
+end
+
+@i function idot(r::T, A::SparseMatrixCSC{T},B::SparseMatrixCSC{T}) where {T}
+    @routine @invcheckoff begin
+        (m, n) ← size(A)
+        branch_keeper ← zeros(Bool, 2*m)
+    end
+    @safe size(B) == (m,n) || throw(DimensionMismatch("matrices must have the same dimensions"))
+    @invcheckoff @inbounds for j = 1:n
+        @routine begin
+            ia1 ← A.colptr[j]
+            ib1 ← B.colptr[j]
+            ia2 ← A.colptr[j+1]
+            ib2 ← B.colptr[j+1]
+            ia ← ia1
+            ib ← ib1
+        end
+        @inbounds for i=1:ia2-ia1+ib2-ib1-1
```

Comment:
@GiggleLiu will this introduce implicit memory allocation?

Reply:
No, its inverse is
```julia
var = ia2-ia1+ib2-ib1-1
@inbounds for i=ia2-ia1+ib2-ib1-1:-1:1
    ...
end
@assert var == ia2-ia1+ib2-ib1-1
```
The `var` is "allocated" on register. In addition, if you put a `@invcheckoff` before this statement, the `@assert` statement will be removed.